src/test/fuzz/chainstate.cpp

```diff
@@ -0,0 +1,699 @@
+// Copyright (c) 2023 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <chain.h>
+#include <chainparams.h>
+#include <config/bitcoin-config.h>
+#include <consensus/merkle.h>
+#include <kernel/notifications_interface.h>
+#include <node/blockstorage.h>
+#include <node/chainstate.h>
+#include <node/miner.h>
+#include <pow.h>
+#include <random.h>
+#include <scheduler.h>
+#include <undo.h>
+#include <validation.h>
+#include <validationinterface.h>
+#include <test/fuzz/FuzzedDataProvider.h>
+#include <test/fuzz/fuzz.h>
+#include <test/fuzz/util.h>
+#include <test/util/mining.h>
+#include <test/util/script.h>
+#include <test/util/setup_common.h>
+#include <txdb.h>
+#include <util/fs_helpers.h>
+#include <util/thread.h>
+
+#include <sys/mman.h>
+
+
+namespace {
+
+const BasicTestingSetup* g_setup;
+
+class KernelNotifications : public kernel::Notifications
+{
+public:
+    kernel::InterruptResult blockTip(SynchronizationState, CBlockIndex&) override { return {}; }
+    void headerTip(SynchronizationState, int64_t height, int64_t timestamp, bool presync) override {}
+    void progress(const bilingual_str& title, int progress_percent, bool resume_possible) override {}
+    virtual void warningSet(kernel::Warning id, const bilingual_str& message) override {}
+    virtual void warningUnset(kernel::Warning id) override {}
+    void flushError(const bilingual_str& debug_message) override
+    {
+        assert(false);
+    }
+    void fatalError(const bilingual_str& message) override
+    {
+        assert(false);
+    }
+};
+
+class DummyQueue : public util::TaskRunnerInterface
+{
+public:
+    DummyQueue() {}
+
+    void insert(std::function<void()> func) override {}
+
+    void flush() override {}
+
+    size_t size() override { return 0; }
+};
+
+//! See net_processing.
+static const int MAX_HEADERS_RESULTS{2000};
+
+//! To generate a random tmp datadir per process (necessary to fuzz with multiple cores).
+static FastRandomContext g_insecure_rand_ctx_temp_path;
+
+struct TestData {
+    fs::path init_datadir;
+    fs::path working_datadir;
+    ValidationSignals main_signals{std::make_unique<DummyQueue>()};
+    KernelNotifications notifs;
+
+    void Init() {
+        const auto rand_str{g_insecure_rand_ctx_temp_path.rand256().ToString()};
+        const auto tmp_dir{fs::temp_directory_path() / "fuzz_chainstate_" PACKAGE_NAME / rand_str};
+        init_datadir = tmp_dir / "init";
+        fs::remove_all(init_datadir);
+        fs::create_directories(init_datadir / "blocks");
+        working_datadir = tmp_dir / "working";
+    }
+
+    ~TestData() {
+        fs::remove_all(init_datadir);
```

Comment:
This should nuke the whole parent tmp dir that was created for the fuzz test, not just `init_datadir`. I'm seeing tons of left over directories otherwise.

Reply:
Done.